src/fabric/src/fabric2_encryption.erl

```diff
@@ -0,0 +1,254 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(fabric2_encryption).
+-behaviour(gen_server).
+-vsn(1).
+
+
+-export([
+    start_link/0,
+    get_wrapped_kek/1,
+    encode/5,
+    decode/5
+]).
+
+
+-export([
+    init/1,
+    terminate/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3
+]).
+
+
+-export([
+    do_encode/6,
+    do_decode/6
+]).
+
+
+-define(INIT_TIMEOUT, 60000).
+-define(LABEL, "couchdb-aes256-gcm-encryption-key").
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+get_wrapped_kek(DbName) when is_binary(DbName) ->
+    case config:get_boolean("encryption", "enabled", false) of
+        true -> gen_server:call(?MODULE, {get_wrapped_kek, DbName});
+        false -> {ok, false}
+    end.
+
+
+encode(WrappedKEK, DbName, DocId, UpdateCounter, DocBody)
+    when is_binary(WrappedKEK),
+         is_binary(DbName),
+         is_binary(DocId),
+         is_integer(UpdateCounter), UpdateCounter > 0,
+         is_binary(DocBody) ->
+    gen_server:call(?MODULE,
```

Comment:
doing all the crypto work in a single gen_server sounds like a recipe for performance issues, even if we do spawn off worker processes there.

could we instead a) set the sensitive flag on self(), b) fetch the unwrapped key c) perform the operation in process d) unset sensitive flag?

Reply:
This is because of KEK cache. Its erts table set to `protected`, so only gen server can read it, fetching unwrapped key from worker would require to change ets to `private`, but then any proc could read from it, there are no option to limit this, unfortunally.

Reply:
you mean `private` there. a `protected` table can be read by all processes.

in the current code you hand the unwrapped key to another process (`spawn_monitor`). It doesn't appear to be more protected in that process than if we did the work in the calling process.

Reply:
yes, `private`, thank you. it's more protected in sense that I can't remsh and read unwrapped key from ets. if you think this is not a concern then I'll make worker to do the unwrap fetch.